storages/backends/gcloud.py

```diff
@@ -141,11 +144,20 @@ def get_default_settings(self):
             # roll over.
             "max_memory_size": setting("GS_MAX_MEMORY_SIZE", 0),
             "blob_chunk_size": setting("GS_BLOB_CHUNK_SIZE"),
+            "sa_email": setting("GS_SA_EMAIL")
         }
 
     @property
     def client(self):
         if self._client is None:
+            if self.project_id is None or self.credentials is None:
+                self.credentials, self.project_id = auth.default(
+                    scopes=['https://www.googleapis.com/auth/cloud-platform']
+                )
+                if not self.credentials.token_state == TokenState.FRESH:
+                    self.credentials.refresh(requests.Request())
```

Comment:
This is inside the cache check, does it need to be moved outside? Something like (e.g)

```python
    @property
    def client(self):
        if self._client is None:
            if self.project_id is None or self.credentials is None:
                self.credentials, self.project_id = auth.default(
                    scopes=['https://www.googleapis.com/auth/cloud-platform']
                )
                if not hasattr(self.credentials, "service_account_email") and self.sa_email:
                    self.credentials.service_account_email = self.sa_email
            self._client = Client(project=self.project_id, credentials=self.credentials)

        if self.credentials and self.credentials.token_state != TokenState.FRESH:
            self.credentials.refresh(requests.Request())
        return self._client
```

Reply:
adjusted for this as well